### PR TITLE
feat(app): Add upload protocol warning modal

### LINKED ISSUE #1032: Usability: Warn User Before Protocol Upload
As a Run App user, I would like to be alerted that uploading a new protocol will clear previous protocol data.

## Acceptance Criteria
- User shown new dialogue box when they attempt to upload a new protocol
- Dialogue box alerts user that opening a new protocol will clear current protocol data and calibration
- User has option to cancel or continue with upload
- Remove current warning message from below the upload feature
 
## Design
- Zeplin > Run App Screens > "open a new file" tag

### PR DESCRIPTION
**Pull request by Kadee80**
## overview

This PR depends on #1969 being merged in first.

Closes #1032

## changelog

- feat(app): Add upload protocol warning modal 

## review requests

Connect to robot, upload a protocol
- [ ] First upload renders and behaves as expected

Upload another protocol (with a session already loaded)
- [ ] Upload warning modal renders
- [ ] Clicking [cancel] returns user to normal upload page, second file is not uploaded
- [ ] Clicking [continue] uploads the correct second protocol

### CODE REVIEW

File: api/opentrons/api/session.py [hunk 1]
```diff
@@ -26,6 +26,11 @@ def create(self, name, text):
         self.session = Session(name=name, text=text)
         return self.session
 
+    def clear(self):
```
Comment (Kadee80) [inline api/opentrons/api/session.py]: This is from rebasing #1969

File: app/src/components/UploadPanel/Upload.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    if (files.length < 1) {
+      return
+    }
+
+    this.setState({uploadedFile: files[0]})
+    this.props.confirmUpload()
+    // $FlowFixMe: type EventTarget missing value
+    event.target.value = null
```
Comment (Kadee80) [inline app/src/components/UploadPanel/Upload.js]: Flow 🔥

File: app/src/components/UploadPanel/Upload.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    if (uploadedFile && !this.props.sessionLoaded) {
+      this.props.createSession(uploadedFile)
+      this.setState({uploadedFile: null})
+    }
+  }
+
+  render () {
+    return (
+      <div>
+        <UploadInput onUpload={this.onUpload} isButton value={null}/>
```
Comment (Kadee80) [inline app/src/components/UploadPanel/Upload.js]: `value={null}` needed to allow for reuploading the same filename open to other suggestions here

File: app/src/components/UploadPanel/Upload.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
+// @flow
+import * as React from 'react'
+import UploadInput from './UploadInput'
+
+type Props = {
+  sessionLoaded: ?boolean,
+  createSession: (file: File) => void,
```
Comment (mcous) [inline app/src/components/UploadPanel/Upload.js]: Same comment as above about `mixed` vs `void`

File: app/src/components/UploadPanel/Upload.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    const {uploadedFile} = this.state
+    if (uploadedFile && !this.props.sessionLoaded) {
+      this.props.createSession(uploadedFile)
+      this.setState({uploadedFile: null})
+    }
+  }
+
+  render () {
+    return (
+      <div>
```
Comment (mcous) [inline app/src/components/UploadPanel/Upload.js]: Should this be a `React.Fragment`?

File: app/src/components/UploadPanel/Upload.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  ) => {
+    let files: Array<File> = []
+    if (event.dataTransfer && event.dataTransfer.files) {
+      files = (event.dataTransfer.files: any)
+    } else if (event.target.files) {
+      files = (event.target.files: any)
+    }
+
+    if (files.length < 1) {
+      return
```
Comment (mcous) [inline app/src/components/UploadPanel/Upload.js]: This early return will prevent the `null`ing out of `event.target.value`, which is necessary to reset input state for repeated uploads

File: app/src/components/UploadPanel/Upload.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    this.setState({uploadedFile: files[0]})
```
Comment (mcous) [inline app/src/components/UploadPanel/Upload.js]: There should be more logic here. We only need to set the state here if there is an existing session. If there is no existing session we should just immediately call `createSession`

```js
if (files.length > 1) {
  if (this.props.sessionLoaded) {
    this.setState({uploadedFile: files[0]})
    this.props.confirmUpload()
  } else {
    this.props.createSession(files[0])
  }
}
```

File: app/src/components/UploadPanel/Upload.js (hunk 3 above)
Comment (mcous) [reply app/src/components/UploadPanel/Upload.js]: If we prefer this, then the `event.target.value = null` is not necessary.

Also, nitpick: missing whitespace before JSX end tag

File: app/src/components/UploadPanel/index.js [hunk 8]
```diff
@@ -1,58 +1,45 @@
...
+import Upload from './Upload'
 
-export default connect(mapStateToProps, mapDispatchToProps)(UploadPanel)
-
-UploadPanel.propTypes = {
-  isSessionLoaded: PropTypes.bool.isRequired,
-  onUpload: PropTypes.func.isRequired
+type Props = {
+  sessionLoaded: ?boolean,
+  confirmUpload: () => void,
```
Comment (mcous) [inline app/src/components/UploadPanel/index.js]: `s/void/mixed`

File: app/src/components/UploadPanel/index.js [hunk 9]
```diff
@@ -1,58 +1,45 @@
...
-        files = event.dataTransfer.files
-      } else {
-        files = event.target.files
-      }
-
-      dispatch(robotActions.session(files[0]))
-
-      // reset the state of the input to allow file re-uploads
-      event.target.value = null
+    confirmUpload: () => {
```
Comment (mcous) [inline app/src/components/UploadPanel/index.js]: No curly braces needed if return type is `mixed`

File: app/src/pages/Upload.js [hunk 10]
```diff
@@ -1,12 +1,22 @@
...
+type Props = {
+  match: Match
+}
+
+export default function UploadPage (props: Props) {
+  const {match: {path}} = props
   return (
     <Page>
       <UploadStatus />
+      <Switch>
```
Comment (mcous) [inline app/src/pages/Upload.js]: Switch is not necessary if we only have this one route

File: app/src/robot/actions.js [hunk 11]
```diff
@@ -299,6 +307,10 @@ export const actions = {
     }
   },
 
+  clearSession () {
```
Comment (mcous) [inline app/src/robot/actions.js]: I don't see any additions to the RPC API client to handle this action and call `session_manager.clear`

Comment (Kadee80) [reply app/src/robot/actions.js]: Good catch

**Pull request merged**